### PR TITLE
Run app in production mode

### PR DESCRIPTION
**Pull request by tpinney**
### Context

Up to this stage all deployments and testing in the Azure CI/CD pipeline was performed with RAILS_ENV set to 'development'. This change reworks the docker image and pipelines so that production and test environments are used in the appropriate places instead.

### Changes proposed in this pull request

- Reworked docker image to include pre-compiled assets.
- Revised azure pipeline files to include appropriate RAILS_ENV declarations where required.
- Reworked environment variables to enable appropriate use of environment variables in the correct places.
- Mapped the Rails SECRET_KEY_BASE variable to a variable called railsSecretKeyBase within the Azure pipeline due to the use of a protected variable name within the azure pipelines.
- Added a new variable under Microsofts guidance to resolve the issue of continued intermittent swap slot failures observed during application deployment.

### Guidance to review

Besides the changes to the Azure pipeline code, I have required changes to some of the application code that I would appreciate particular attention on in case I have unwittingly introduced unexpected issues there.

Files of note include:
- Dockerfile
- Makefile
- config/database.yml
- config/environments/development.yml
- config/environments/production.yml

### Link to Trello card

[772 - Run the app in production mode on azure](https://trello.com/c/YovhLqlr/772-run-the-app-in-production-mode-on-azure)

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -38,41 +38,49 @@ stages:
...
         make ci.test
+        test_result=$?
+        if [ "$test_result" == "0" ]; then true; else cat rspec-results.xml; false; fi
       displayName: 'Execute tests'
       env:
         dockerHubUsername: $(dockerHubUsername)
         dockerHubImageName: $(imageName)
-        GOVUK_NOTIFY_API_KEY: $(govukNotifyAPIKey)
+        railsSecretKeyBase: $(railsSecretKeyBase)
+        RAILS_ENV: test
```
Comment (tijmenb) [inline azure-pipelines.yml]:
```suggestion
        RAILS_ENV: production
```

File: Dockerfile [hunk 2]
```diff
@@ -1,19 +1,19 @@
...
-RUN yarn install --check-files
-RUN bundle install
+RUN gem install bundler:2.0.2 && \
+    yarn install --check-files && \
+    bundle install
 
-ADD . $APP_HOME/
+RUN bundle exec rake assets:precompile
 
-CMD bundle exec rails db:migrate && bundle exec rails server -b 0.0.0.0
```
Comment (tijmenb) [inline Dockerfile]: Why are we no longer migrating the database? Will that have to be done manually?

File: Makefile [hunk 3]
```diff
@@ -10,7 +10,6 @@ build: ## Create a new image
 setup: build ## Set up a clean database and node_modules folder for running the app or the specs in docker
 	docker-compose down -v
 	docker-compose run --rm web bundle exec rake db:setup
-	docker-compose run --rm web yarn install
```
Comment (tijmenb) [inline Makefile]: Is this step no longer necessary?

File: azure-pipelines.yml [hunk 4]
```diff
@@ -38,41 +38,49 @@ stages:
...
 
-    - script: |
-        make setup
+    - script: make setup
       displayName: 'Build & setup'
       env:
         dockerHubUsername: $(dockerHubUsername)
         dockerHubImageName: $(imageName)
-        GOVUK_NOTIFY_API_KEY: $(govukNotifyAPIKey)
+        railsSecretKeyBase: $(railsSecretKeyBase)
```
Comment (duncanjbrown) [inline azure-pipelines.yml]: 👌 Glad this is working now

File: azure-pipelines.yml [hunk 5]
```diff
@@ -38,41 +38,49 @@ stages:
...
-    - script: |
-        make ci.lint-ruby
+    - script: make ci.lint-ruby
       displayName: 'Rubocop'
       env:
         dockerHubUsername: $(dockerHubUsername)
         dockerHubImageName: $(imageName)
-        GOVUK_NOTIFY_API_KEY: $(govukNotifyAPIKey)
+        railsSecretKeyBase: $(railsSecretKeyBase)
+        RAILS_ENV: production
```
Comment (duncanjbrown) [inline azure-pipelines.yml]: Is it correct to set `production` as the env for linting? Can we get away with `test` here?

File: config/environments/development.rb [hunk 6]
```diff
@@ -36,7 +36,7 @@
   config.action_mailer.perform_caching = false
   config.action_mailer.delivery_method = :notify
   config.action_mailer.notify_settings = {
-    api_key: ENV.fetch('GOVUK_NOTIFY_API_KEY')
+    api_key: ENV['GOVUK_NOTIFY_API_KEY']
```
Comment (duncanjbrown) [inline config/environments/development.rb]: Why this change? `fetch` would throw an exception if the key weren't set

File: config/environments/development.rb [hunk 7]
```diff
@@ -36,7 +36,7 @@
   config.action_mailer.perform_caching = false
   config.action_mailer.delivery_method = :notify
   config.action_mailer.notify_settings = {
-    api_key: ENV.fetch('GOVUK_NOTIFY_API_KEY')
```
Comment (tijmenb) [inline config/environments/development.rb]: Is this change necessary? The upside of `fetch` is that we won't be suprised by nil-errors.

File: azure-pipelines.yml [hunk 8]
```diff
@@ -38,41 +38,49 @@ stages:
...
+        docker-compose run --rm web bundle exec rake db:setup
         make ci.test
+        test_result=$?
+        if [ "$test_result" == "0" ]; then true; else cat rspec-results.xml; false; fi
       displayName: 'Execute tests'
       env:
         dockerHubUsername: $(dockerHubUsername)
         dockerHubImageName: $(imageName)
-        GOVUK_NOTIFY_API_KEY: $(govukNotifyAPIKey)
+        railsSecretKeyBase: $(railsSecretKeyBase)
```
Comment (duncanjbrown) [inline azure-pipelines.yml]: I think we can lose secretKeyBase if we're not in `RAILS_ENV=production`

File: azure-pipelines.yml [hunk 9]
```diff
@@ -38,41 +38,49 @@ stages:
...
-    - script: |
-        make setup
+    - script: make setup
       displayName: 'Build & setup'
       env:
         dockerHubUsername: $(dockerHubUsername)
         dockerHubImageName: $(imageName)
-        GOVUK_NOTIFY_API_KEY: $(govukNotifyAPIKey)
+        railsSecretKeyBase: $(railsSecretKeyBase)
+        RAILS_ENV: production
```
Comment (duncanjbrown) [inline azure-pipelines.yml]: How do we create the test db here if RAILS_ENV is `production`? Going to have a look at this locally

File: config/environments/development.rb (hunk 7 above)
Comment (tpinney) [reply config/environments/development.rb]: This is an awkward one because during the image build I you get an error about this key not being present in the image, which would be a security risk if it were present in the image.
The only way I could see for that not to raise an error during the build was to remove that fetch, but happy to change this if anyone can suggest an alternative.

File: azure-pipelines.yml (hunk 9 above)
Comment (tpinney) [reply azure-pipelines.yml]: There is a step in the azure pipeline to run the db:setup again in the container that used during build.

File: azure-pipelines.yml (hunk 8 above)
Comment (tpinney) [reply azure-pipelines.yml]: I can double check, but I'm pretty sure I got an error that the key wasn't present. I only seemed to be able to run it without that key being present in dev mode.

File: azure-pipelines.yml (hunk 5 above)
Comment (tpinney) [reply azure-pipelines.yml]: I can change it to test, but seemed happy to run the tests in production mode for that if you want me to make that change.

File: config/environments/development.rb (hunk 6 above)
Comment (tpinney) [reply config/environments/development.rb]: See my comment to @tijmenb above. He raised the same query.

File: Dockerfile (hunk 2 above)
Comment (duncanjbrown) [reply Dockerfile]: +1

File: Dockerfile [hunk 10]
```diff
@@ -1,19 +1,19 @@
...
+CMD rm -f tmp/pids/server.pid && bundle exec rails server -b 0.0.0.0
```
Comment (duncanjbrown) [inline Dockerfile]: This is already in the docker-compose entrypoint for the `web` service: https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/blob/master/docker-compose.yml#L25

File: Makefile (hunk 3 above)
Comment (tpinney) [reply Makefile]: The assets:precompile step in the Dockerfil triggers the install steps so this was was introducing an unnecessary delay doing it again here, so I removed it.

File: azure-pipelines.yml (hunk 8 above)
Comment (tijmenb) [reply azure-pipelines.yml]: Rails should generate a random secret in test and development:

https://github.com/rails/rails/blob/516474def169b38caa743eb44a87e86a12de0595/railties/lib/rails/application.rb#L421-L427

File: Makefile (hunk 3 above)
Comment (duncanjbrown) [reply Makefile]: I'm not sure why it shouldn't be — `node_modules` is a volume so on a freshly created setup it will be empty?

File: azure-pipelines.yml (hunk 5 above)
Comment (tpinney) [reply azure-pipelines.yml]: I've changed this to test mode.

File: azure-pipelines.yml (hunk 8 above)
Comment (tpinney) [reply azure-pipelines.yml]: Having re-run this in test mode without the rails secret key environment variable it appears to be working. Not sure what was difference when it failed before.

File: Dockerfile (hunk 10 above)
Comment (duncanjbrown) [reply Dockerfile]: See 528fe342

File: Makefile (hunk 3 above)
Comment (duncanjbrown) [reply Makefile]: The `yarn install` in the Dockerfile will bake `node_modules` into the container, and I agree this is what we want for an production image.

When the app is run via `docker-compose.yml` a docker volume is mounted for 
`node_modules`. (see the `volumes` key in that file). This is to support local development workflows where the app may be run outside Docker and the host's `node_modules` folder may clash with the one inside (see b1b28fc1).

I was under the impression that this meant starting the app from cold meant the `node_modules` volume was empty, but reading up, I didn't understand `docker-compose` volume behaviour. From the [docs](https://docs.docker.com/storage/volumes/):

> If you start a container which creates a new volume... and the container has files or directories in the directory to be mounted (such as /app/ above), the directory’s contents are copied into the volume. The container then mounts and uses the volume, and other containers which use the volume also have access to the pre-populated content.

**tl;dr** this change is :+1:

Comment (duncanjbrown) [reply Makefile]: We discussed this IRL and agreed that a stock `docker-compose.override.yml` holding dev-only config would clarify things somewhat. https://docs.docker.com/compose/extends/#adding-and-overriding-configuration. In this configuration we don't mount node_modules (or `/app` 🎉 ) during the build process in Azure

File: Dockerfile (hunk 2 above)
Comment (tpinney) [reply Dockerfile]: Addressed in commit 86d4973

File: Makefile (hunk 3 above)
Comment (tpinney) [reply Makefile]: Addressed as per discussion with @duncanjbrown in commit 86d4973.

File: Dockerfile (hunk 10 above)
Comment (tpinney) [reply Dockerfile]: Addressed in 86d4973.

File: Dockerfile [hunk 11]
```diff
@@ -1,19 +1,19 @@
...
+
 RUN apk add --update build-base postgresql-dev git tzdata nodejs yarn && \
     cp /usr/share/zoneinfo/Europe/London /etc/localtime && \
     echo "Europe/London" > /etc/timezone
 
-RUN gem install bundler:2.0.2
-
-ENV APP_HOME /app
-RUN mkdir $APP_HOME
+COPY . $APP_HOME/
```
[the remark this answers is not in the document]
Comment (tpinney) [reply Dockerfile]: Addressed in 86d4973.

File: config/environments/development.rb (hunk 7 above)
Comment (tpinney) [reply config/environments/development.rb]: Having discussed this with @duncanjbrown  I will see if I can temporarily set a dummy value for the GOVUK_NOTIFY_API_KEY in the image generated in the Dockerfile and then successfully override that with an environment variable through docker-compose in the runtime container.
This seems to be the best way to work around this issue for now, as opposed to my previous solution of not using fetch.

Comment (tpinney) [reply config/environments/development.rb]: Setting a dummy value for the GOVUK_NOTIFY_API_KEY in the Dockerfile works.
Commit 258f6e9 addresses this.

**Comment by tpinney** [comment]
@duncanjbrown & @tijmenb I believe all comments have now been addressed. The build has just passed and its currently deploying into Azure. Can you please review again.

If you could hold fire on merging if you're happy because I'll need to reinstate the condition on the dev deployment to only happen on the master branch and not all branches.

**Comment by tpinney** [comment]
> @duncanjbrown & @tijmenb I believe all comments have now been addressed. The build has just passed and its currently deploying into Azure. Can you please review again.
> 
> If you could hold fire on merging if you're happy because I'll need to reinstate the condition on the dev deployment to only happen on the master branch and not all branches.

I have restored the dev deployment conditions for master branch only, so this change is now safe to merge if everyone is happy with the amendments I've made.

**Comment by duncanjbrown** [comment]
I have an idea to take out a setup step: will push on here and see if it works

**Comment by duncanjbrown** [comment]
That worked — I think I'm happy — one more look @tijmenb & @tpinney?

**Comment by duncanjbrown** [comment]
We're not minifying our assets in prod because we're not running assets:precompile under `RAILS_ENV=production`. This can be a future PR: https://trello.com/c/WV0XHND1/1013-minify-assets-in-production

**Comment by duncanjbrown** [comment]
I need to fix up some env vars before this is ready

**Comment by duncanjbrown** [comment]
Entering your email and submitting the form results in a 500. However, everything else is working.

To fix that error we need either

- logs from the production container
- a deployment with `config.consider_all_requests_local`

I tried to do the latter — twice — and both deployments failed.

If we're happy with a card to fix the logging and the fix the issue, I think this should be mergeable?
